src/Microsoft.PowerShell.Commands.Utility/commands/utility/Csv.cs

```diff
@@ -27,85 +29,60 @@ internal CSVHelper(char delimiter)
+                char nextChar = (char)reader.Read();
+
+                // if next character was delimiter or we are at the end, add string to result and clear builder
+                // else if next character was quote, perform reading untill next quote and add it to builder
+                // else read and add it to builder
+                if (nextChar == Delimiter) 
                 {
-                    if (!inQuote)
-                    {
-                        result.Add(tempString);
-                        tempString = string.Empty;
-                    }
-                    else
-                    {
-                        tempString += c;
-                    }
-                }
-                else
+                    result.Add(wordBuffer.ToString());
+                    wordBuffer.Clear();
+                } 
+                else if (nextChar == '"') 
                 {
-                    switch (c)
+                    bool inQuotes = true;
+                    while (reader.Peek() != -1 && inQuotes) 
                     {
-                        case '"':
-                            if (inQuote)
-                            {
-                                // If we are at the end of the string or the end of the segment, create a new value
-                                // Otherwise we have an error
-                                if (i == csv.Length - 1)
-                                {
-                                    result.Add(tempString);
-                                    tempString = string.Empty;
-                                    inQuote = false;
-                                    break;
-                                }
-
-                                if (csv[i + 1] == Delimiter)
-                                {
-                                    result.Add(tempString);
-                                    tempString = string.Empty;
-                                    inQuote = false;
-                                    i++;
-                                }
-                                else if (csv[i + 1] == '"')
-                                {
-                                    tempString += '"';
-                                    i++;
-                                }
-                                else
-                                {
-                                    inQuote = false;
-                                }
-                            }
-                            else
-                            {
-                                inQuote = true;
-                            }
-
-                            break;
-
-                        default:
-                            tempString += c;
-                            break;
+                        nextChar = (char)reader.Read();
+                        
+                        if (nextChar == '"') 
+                        {
+                            inQuotes = false;
```

Comment:
You are not handling the input like this: `"abc""def"`. This will happen when the original value contains `"` in it.

Reply:
Sorry for my late reply. 

I'm not entirely sure what you mean. As far as I can see the functionality of the old and new implementation correspond, in the sense that the `"` is excluded in the content. `"abc""def"` should parse to `[abc,def]`. Did you spot an inconsistency with respect to the old implementation or are you pointing at something else?

Reply:
`"abc""def"` will be parsed into `abc"def` in the original code.

Reply:
ah, I was assuming otherwise. Good that you mention this. Apparently this is not covered in the tests, since the build passes.

Reply:
so to be correct: whenever we encounter `""` we should interpret this as a quote character, and when we encounter `"` we should interpret it as an escape quote